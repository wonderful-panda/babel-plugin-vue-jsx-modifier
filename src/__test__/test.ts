import * as babel from "@babel/core";
import plugin from "../..";

function transform(code: string): string | undefined {
  const opts: babel.TransformOptions = {
    plugins: [plugin]
  };
  const ret = babel.transform(code, opts);
  return ret ? ret.code || undefined : undefined;
}

describe("__sync", () => {
  it("basic functionary", () => {
    const code = transform(`<MyComponent foo={ __sync(state.fooValue) } />`);
    expect(code).toMatchSnapshot();
  });

  it("object name of __sync method is just ignored", () => {
    const code = transform(
      `<MyComponent foo={ xxx.__sync(state.fooValue) } />`
    );
    expect(code).toMatchSnapshot();
  });

  it("works for deep member expression", () => {
    const code = transform(`<MyComponent foo={ __sync(state.foo.bar.baz) } />`);
    expect(code).toMatchSnapshot();
  });

  it("works for literal member expression", () => {
    const code = transform(`<MyComponent foo={ __sync(state['foo-key']) } />`);
    expect(code).toMatchSnapshot();
  });

  it("update handlers of multiple sync props are in same object", () => {
    const code = transform(
      `<MyComponent foo={ __sync(state.fooValue) } bar={ __sync(state.barValue) } />`
    );
    expect(code).toMatchSnapshot();
  });

  it("avoid using argument name which shadows existing variables", () => {
    const code = transform(
      `const _v0 = <MyComponent foo={ __sync(state.fooValue) } bar={ __sync(state.barValue) } />;`
    );
    expect(code).toMatchSnapshot();
  });

  it("update handler of kebab-case prop should be camelCase", () => {
    const code = transform(
      `<MyComponent foo-bar-baz={ __sync(state.value ) } />`
    );
    expect(code).toMatch(/"update:fooBarBaz":/);
  });

  it("Throw error if sprecified in event handler", () => {
    expect(() => {
      transform(`<div onClick={ __sync(this.onClick) } />`);
    }).toThrow(/only in component prop/);

    expect(() => {
      transform(`<MyComponent nativeOnClick={ __sync(this.onClick) } />`);
    }).toThrow(/only in component prop/);
  });

  it("Throw error if argument is not MemberExpression", () => {
    expect(() => {
      transform(`<div foo={ __sync(data) } />`);
    }).toThrow(/must be MemberExpression/);
  });
});

describe("__relay", () => {
  it("basic functionary", () => {
    const code = transform(`<MyComponent foo={ __relay(this.fooValue) } />`);
    expect(code).toMatchSnapshot();
  });

  it("works for literal MemberExpression", () => {
    const code = transform(`<MyComponent foo={ __relay(this['foo-key']) } />`);
    expect(code).toMatchSnapshot();
  });

  it("works for computed MemberExpression", () => {
    const code = transform(`<MyComponent foo={ __relay(this[fooKey]) } />`);
    expect(code).toMatchSnapshot();
  });

  it("update handlers of multiple sync/relay props are in same object", () => {
    const code = transform(
      `<MyComponent foo={ __sync(this.fooValue) } bar={ __relay(this.barValue) } />`
    );
    expect(code).toMatchSnapshot();
  });

  it("Throw error if sprecified in event handler", () => {
    expect(() => {
      transform(`<div onClick={ __relay(this.onClick) } />`);
    }).toThrow(/only in component prop/);

    expect(() => {
      transform(`<MyComponent nativeOnClick={ __relay(this.onClick) } />`);
    }).toThrow(/only in component prop/);
  });

  it("Throw error if argument is not MemberExpression", () => {
    expect(() => {
      transform(`<div foo={ __relay(data) } />`);
    }).toThrow(/must be MemberExpression/);
  });
});

describe("__capture", () => {
  it("basic functionary (on)", () => {
    const code = transform(
      `<div onScroll={ __capture(this.onScrollCapture) } />`
    );
    expect(code).toMatchSnapshot();
  });

  it("basic functionary (nativeOn)", () => {
    const code = transform(
      `<MyComponent nativeOnScroll={ __capture(this.onScrollCapture) } />`
    );
    expect(code).toMatchSnapshot();
  });

  it("Throw error if sprecified in other than event handler", () => {
    expect(() => {
      transform(`<div foo={ __capture(this.onClick) } />`);
    }).toThrow(/only in event handler/);
  });

  it("Event name conversion (onFooBar -> fooBar)", () => {
    const code = transform(`<div onFooBar={ __capture(handler) } />`);
    expect(code).toMatch(/"!fooBar": /);
  });

  it("Event name conversion (nativeOnFooBar -> fooBar)", () => {
    const code = transform(`<div nativeOnFooBar={ __capture(handler) } />`);
    expect(code).toMatch(/"!fooBar": /);
  });

  it("Event name conversion (on-foo-bar -> foo-bar)", () => {
    const code = transform(`<div on-foo-bar={ __capture(handler) } />`);
    expect(code).toMatch(/"!foo-bar": /);
  });

  it("Event name conversion (nativeOn-foo-bar -> foo-bar)", () => {
    const code = transform(`<div nativeOn-foo-bar={ __capture(handler) } />`);
    expect(code).toMatch(/"!foo-bar": /);
  });
});

describe("__passive", () => {
  it("basic functionary", () => {
    const code = transform(
      `<div onScroll={ __passive(this.onScrollCapture) } />`
    );
    expect(code).toMatch(/"&scroll":/);
  });
});

describe("__once", () => {
  it("basic functionary", () => {
    const code = transform(`<div onScroll={ __once(this.onScrollCapture) } />`);
    expect(code).toMatch(/"~scroll":/);
  });
});

describe("__captureOnce", () => {
  it("basic functionary", () => {
    const code = transform(
      `<div onScroll={ __captureOnce(this.onScrollCapture) } />`
    );
    expect(code).toMatch(/"~!scroll":/);
  });
});
