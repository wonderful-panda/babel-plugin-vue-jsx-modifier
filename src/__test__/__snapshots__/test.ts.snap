// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__capture basic functionary (nativeOn) 1`] = `
"<MyComponent {...{
  nativeOn: {
    '!scroll': this.onScrollCapture
  }
}} />;"
`;

exports[`__capture basic functionary (on) 1`] = `
"<div {...{
  on: {
    '!scroll': this.onScrollCapture
  }
}} />;"
`;

exports[`__sync avoid using argument name which shadows existing variables 1`] = `
"const _v0 = <MyComponent foo={state.fooValue} bar={state.barValue} {...{
  on: {
    'update:foo': _v1 => {
      state.fooValue = _v1;
    },
    'update:bar': _v1 => {
      state.barValue = _v1;
    }
  }
}} />;"
`;

exports[`__sync basic functionary 1`] = `
"<MyComponent foo={state.fooValue} {...{
  on: {
    'update:foo': _v0 => {
      state.fooValue = _v0;
    }
  }
}} />;"
`;

exports[`__sync object name of __sync method is just ignored 1`] = `
"<MyComponent foo={state.fooValue} {...{
  on: {
    'update:foo': _v0 => {
      state.fooValue = _v0;
    }
  }
}} />;"
`;

exports[`__sync update handlers of multiple sync props are in same object 1`] = `
"<MyComponent foo={state.fooValue} bar={state.barValue} {...{
  on: {
    'update:foo': _v0 => {
      state.fooValue = _v0;
    },
    'update:bar': _v0 => {
      state.barValue = _v0;
    }
  }
}} />;"
`;

exports[`__sync works for deep member expression 1`] = `
"<MyComponent foo={state.foo.bar.baz} {...{
  on: {
    'update:foo': _v0 => {
      state.foo.bar.baz = _v0;
    }
  }
}} />;"
`;

exports[`__sync works for literal member expression 1`] = `
"<MyComponent foo={state['foo-key']} {...{
  on: {
    'update:foo': _v0 => {
      state['foo-key'] = _v0;
    }
  }
}} />;"
`;
